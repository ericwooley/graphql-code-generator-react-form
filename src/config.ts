import { RawClientSideBasePluginConfig } from '@graphql-codegen/visitor-plugin-common';

/**
 * @description This plugin generates React forms forms.
 *
 * It extends the basic TypeScript plugins: `@graphql-codegen/typescript`, `@graphql-codegen/typescript-operations` - and thus shares a similar configuration.
 */
export interface ReactformsRawPluginConfig
  extends RawClientSideBasePluginConfig {}
