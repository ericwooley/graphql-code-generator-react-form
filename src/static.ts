export const validationFNSource = `
export interface IGenericFormValidationResult {
  __meta?:string;
  [key: string]:
    | undefined
    | string
    | IGenericFormValidationResult
    | (IGenericFormValidationResult | string)[];
}
export const isValidFromFormResult = (
  obj: IGenericFormValidationResult
): boolean => {
  return Object.values(obj).reduce((isValid: boolean, val) => {
    if (!isValid) return false;
    if (val === undefined) return true;
    if (typeof val === 'string') return val.length === 0;
    if (Array.isArray(val))
      return (
        val.find((nestedVal) =>
          typeof nestedVal === 'string'
            ? nestedVal
            : !isValidFromFormResult(nestedVal)
        ) === undefined
      );
    return isValidFromFormResult(val);
  }, true);
};
`;
