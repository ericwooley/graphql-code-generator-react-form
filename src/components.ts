import { pascalCase } from 'change-case-all';
import { ReactFormsRawPluginConfig } from './config';

export class ComponentComposer {
  constructor(protected rawConfig: ReactFormsRawPluginConfig) {}

  private generateComponentRenderer(name: string) {
    const tagName = `${pascalCase(name)}Component`;
    return {
      init: `const ${tagName} = useCustomizedComponent(${JSON.stringify(
        name
      )})`,
      render: (props: { [key: string]: string }, children: string) => {
        return `<${tagName} ${Object.entries(props)
          .map(([key, val]) => `${key}={${val}}`)
          .join(' ')} path={path} scalar={scalar} name={name} ${
          props.depth ? '' : 'depth={depth}'
        }>
          ${children}
          </${tagName}>`;
      },
      tagName,
    };
  }
  get form() {
    return this.generateComponentRenderer('form');
  }
  get submitButton() {
    return this.generateComponentRenderer('submitButton');
  }
  get div() {
    return this.generateComponentRenderer('div');
  }
  get listItem() {
    return this.generateComponentRenderer('listItem');
  }
  get button() {
    return this.generateComponentRenderer('button');
  }
  get label() {
    return this.generateComponentRenderer('label');
  }
  get listWrapper() {
    return this.generateComponentRenderer('listWrapper');
  }
  get labelTextWrapper() {
    return this.generateComponentRenderer('labelTextWrapper');
  }
  get input() {
    return this.generateComponentRenderer('input');
  }
  get addButton() {
    return this.generateComponentRenderer('addButton');
  }
  get removeButton() {
    return this.generateComponentRenderer('removeButton');
  }
  private generatePassthroughComponent(
    name: string,
    componentType = name,
    {
      extra = '',
      propTypes = 'GQLReactFormStandardProps',
    }: { extra?: string; propTypes?: string } = {}
  ) {
    return `${name}: ({...props}: ${propTypes}) => <${componentType} {...props} ${extra}/>`;
  }
  public generateContext(typeComponentMap: { [key: string]: any }) {
    const dynamicComponentTypeList = Object.keys(typeComponentMap)
      .filter(
        (name) =>
          ![
            'StringFormInput',
            'IntFormInput',
            'FloatFormInput',
            'BooleanFormInput',
            'IDFormInput',
          ].includes(name)
      )
      .filter((name) => !name.match(/AsList$/));
    return `
      export interface GQLReactFormStandardProps {
        children?: React.ReactNode
        /**
         * path of properties to get to this item, separated by a dot
         * @example
         * // path = 'UserInput.friends.id'
         * const childOfUser = Boolean(path.split('.').includes('UserInput'))
         **/
        path: string
        id?: string
        /**
         * Number of Scalar Objects deep this property is
         **/
        depth: number,
        className?: string
        name: string
        scalar: string
      }
      export interface GQLReactFormButtonProps extends GQLReactFormStandardProps {
        onClick?: (e?: {preventDefault: () => any}) => any
      }
      interface FormPrimeInput extends React.FC<{onChange: (value: number|string) => number|string, value?: number|string, label: string, onBlur: () => unknown, touched: boolean, error?:string} & GQLReactFormStandardProps> {}
      export interface GQLFormStandardComponent<T extends GQLReactFormStandardProps = GQLReactFormStandardProps> extends React.FC<T> { }
      export interface GQLReactFormListItemProps extends GQLReactFormStandardProps {
        removeButton: JSX.Element|JSX.Element[]
        insertAboveButton: JSX.Element|JSX.Element[]
        idx: number
      }
      export interface GQLReactFormListWrapperProps extends GQLReactFormStandardProps {
        // idx: number
        totalInList: number
      }

      export interface GQLReactFormContext {
        form: GQLFormStandardComponent< GQLReactFormStandardProps&{onSubmit: (e?: {preventDefault?: () => any}) => any} >,
        div: GQLFormStandardComponent,
        label: GQLFormStandardComponent,
        labelTextWrapper: GQLFormStandardComponent,
        button: GQLFormStandardComponent<GQLReactFormButtonProps>,
        addButton: GQLFormStandardComponent<GQLReactFormButtonProps>,
        removeButton: GQLFormStandardComponent<GQLReactFormButtonProps>,
        listWrapper: GQLFormStandardComponent<GQLReactFormListWrapperProps>,
        listItem: GQLFormStandardComponent<GQLReactFormListItemProps>,
        submitButton: React.FC<{text: string}>
        input: FormPrimeInput,
        ${dynamicComponentTypeList
          .map(
            (scalarName) =>
              `${scalarName.replace(
                /FormInput$/,
                ''
              )}?: React.FC<${scalarName}PropTypes> `
          )
          .join(',\n')}
      }
      export const defaultReactFormContext: GQLReactFormContext = {
        form: 'form' as any,
        ${this.generatePassthroughComponent('div')},
        ${this.generatePassthroughComponent('label')},
        ${this.generatePassthroughComponent('labelTextWrapper', 'h4')},
        ${this.generatePassthroughComponent('button', 'button', {
          extra: 'onClick={e => {e.preventDefault(); props.onClick?.()}}',
          propTypes: 'GQLReactFormButtonProps',
        })},
        ${this.generatePassthroughComponent('addButton', 'button', {
          extra: 'onClick={e => {e.preventDefault(); props.onClick?.()}}',
          propTypes: 'GQLReactFormButtonProps',
        })},
        ${this.generatePassthroughComponent('removeButton', 'button', {
          extra: 'onClick={e => {e.preventDefault(); props.onClick?.()}}',
          propTypes: 'GQLReactFormButtonProps',
        })},
        ${this.generatePassthroughComponent('listWrapper', 'ol')},
        listItem: (props: GQLReactFormListItemProps) => {
          return (
            <li>
              {props.insertAboveButton}
              {props.removeButton}
              {props.children}
            </li>
          )
        },
        submitButton: ((props: {text: string}) => <input type="submit" {...props} value={props.text} /> ),
        input: ((props) => {
          const {path, scalar, name, depth, error} = props
          ${this.div.init}
          ${this.labelTextWrapper.init}
          const typeofValue = typeof props.value
          ${this.label.init}
            return (
            ${this.div.render(
              {},
              `
            ${this.label.render(
              {},
              `
              ${this.labelTextWrapper.render({}, `{props.label}`)}
              <><input value={props.value} type={props.value === undefined
                ? "string"
                : typeofValue === 'string'
                ? 'text'
                : typeofValue === 'number'
                ? 'number'
                : ''
              } onChange={(e) =>
                props.onChange(e.target.value)} />
                {error}
              </>
            `
            )}
            `
            )})

          }),

      }
      export const GQLReactFormContext = React.createContext<Partial<GQLReactFormContext>>(defaultReactFormContext)
      const _emptyFormContext = {};
      export const InternalGQLReactFormContext = React.createContext<Partial<GQLReactFormContext>>(_emptyFormContext)
      function useCustomizedComponent <T extends keyof GQLReactFormContext>(name: T): GQLReactFormContext[T] {
        const ctx = React.useContext(GQLReactFormContext)
        const _ctx = React.useContext(InternalGQLReactFormContext)
        let c: GQLReactFormContext[T] = (_ctx[name] || ctx[name] || defaultReactFormContext[name]) as GQLReactFormContext[T]
        return c
      }

    `;
  }
}
