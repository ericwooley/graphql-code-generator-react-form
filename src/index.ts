import {
  Types,
  PluginValidateFn,
  PluginFunction,
} from '@graphql-codegen/plugin-helpers';

import {
  visit,
  GraphQLSchema,
  concatAST,
  Kind,
  FragmentDefinitionNode,
} from 'graphql';
import { LoadedFragment } from '@graphql-codegen/visitor-plugin-common';
import { ReactFormsVisitor } from './visitor';
import { ReactFormsRawPluginConfig } from './config';
import { extname } from 'path';

export const plugin: PluginFunction<
  ReactFormsRawPluginConfig,
  Types.ComplexPluginOutput
> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: ReactFormsRawPluginConfig
) => {
  const allAst = concatAST(documents.map((v) => v.document as any));

  const allFragments: LoadedFragment[] = [
    ...(allAst.definitions.filter(
      (d) => d.kind === Kind.FRAGMENT_DEFINITION
    ) as FragmentDefinitionNode[]).map((fragmentDef) => ({
      node: fragmentDef,
      name: fragmentDef.name.value,
      onType: fragmentDef.typeCondition.name.value,
      isExternal: false,
    })),
    ...(config.externalFragments || []),
  ];

  const visitor = new ReactFormsVisitor(
    schema,
    allFragments,
    config,
    documents
  );

  visit(allAst, { leave: visitor });
  visitor.generateFormsOutput();
  return {
    prepend: visitor.formsImports(),
    content: [visitor.sdkContent].join('\n'),
  };
};

export const validate: PluginValidateFn<any> = async (
  _schema: GraphQLSchema,
  _documents: Types.DocumentFile[],
  _config: ReactFormsRawPluginConfig,
  outputFile: string
) => {
  if (extname(outputFile) !== '.tsx') {
    throw new Error(
      `Plugin "typescript-react-forms" requires extension to be ".tsx"`
    );
  }
};

export { ReactFormsVisitor as ReactApolloVisitor };
