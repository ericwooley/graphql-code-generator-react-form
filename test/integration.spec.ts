import { validateTs } from './typescript-testing-utils';
import { plugin } from '../src/index';
import { parse, GraphQLSchema, buildSchema } from 'graphql';
import { Types, mergeOutputs } from '@graphql-codegen/plugin-helpers';
import { plugin as tsPlugin } from '@graphql-codegen/typescript';
import { plugin as tsDocumentsPlugin } from '@graphql-codegen/typescript-operations';
import { readFileSync } from 'fs';
import { join } from 'path';

describe('React Apollo', () => {
  let spyConsoleError: jest.SpyInstance;
  beforeEach(() => {
    spyConsoleError = jest.spyOn(console, 'warn');
    spyConsoleError.mockImplementation();
  });

  afterEach(() => {
    spyConsoleError.mockRestore();
  });

  const schema = buildSchema(
    readFileSync(
      join(__dirname, '../examples/schemas/schema.graphql')
    ).toString()
  );
  const basicDoc = parse(/* GraphQL */ `
    query users {
      allUsers {
        id
      }
    }
  `);
  const mutationDoc = parse(/* GraphQL */ `
    mutation addUsers($users: [UserInput]!) {
      addUsers(users: $users) {
        id
      }
    }
  `);

  const validateTypeScript = async (
    output: Types.PluginOutput,
    testSchema: GraphQLSchema,
    documents: Types.DocumentFile[],
    config: any
  ) => {
    const tsOutput = await tsPlugin(testSchema, documents, config, {
      outputFile: '',
    });
    const tsDocumentsOutput = await tsDocumentsPlugin(
      testSchema,
      documents,
      config,
      { outputFile: '' }
    );
    const merged = mergeOutputs([tsOutput, tsDocumentsOutput, output]);
    // process.stdout.write(`\n\n---------- merged \n${merged}\n`);
    validateTs(merged, undefined, true, false, [
      `Cannot find namespace 'Types'.`,
    ]);

    return merged;
  };

  describe('Imports', () => {
    it('should import React and forms dependencies', async () => {
      const docs = [{ location: '', document: basicDoc }];
      const content = (await plugin(
        schema,
        docs,
        {},
        {
          outputFile: 'graphql.tsx',
        }
      )) as Types.ComplexPluginOutput;

      expect(content.prepend).toContain(`import * as React from 'react';`);
      await validateTypeScript(content, schema, docs, {});
    });
  });
  describe('Forms', () => {
    it('should generate a Form', async () => {
      const docs = [{ location: '', document: mutationDoc }];
      const content = (await plugin(
        schema,
        docs,
        {},
        {
          outputFile: 'graphql.tsx',
        }
      )) as Types.ComplexPluginOutput;
      expect(content.content).toMatchSnapshot();
      await validateTypeScript(content, schema, docs, {});
    });
  });
});
