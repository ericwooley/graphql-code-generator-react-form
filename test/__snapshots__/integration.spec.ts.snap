// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React Apollo Forms should generate a Form 1`] = `
"
    /**************************
     * utilities
     *************************/
    let idNonce = 0;
    const uniqueId = (inStr: string) => inStr+(idNonce++)
    

    
      interface StandardProps {
        children?: React.ReactNode
        path: string
        id?: string
        className?: string
        name: string
        scalar: string
      }
      interface FormPrimeInput extends React.FC<{onChange: (value: number|string) => number|string, value?: number|string, label: string} & StandardProps> {}
      interface GQLFormStandardComponent<T = {}> extends React.FC<StandardProps & T> { }
      export interface GQLReactFormContext {
        form: GQLFormStandardComponent< {onSubmit: (e?: {preventDefault?: () => any}) => any} >,
        div: GQLFormStandardComponent,
        label: GQLFormStandardComponent,
        labelTextWrapper: GQLFormStandardComponent,
        button: GQLFormStandardComponent<{onClick?: (e?: {preventDefault?: () => any}) => any }>,
        listWrapper: GQLFormStandardComponent,
        listItem: GQLFormStandardComponent,
        submitButton: React.FC<{text: string}>
        input: FormPrimeInput,
        UserInput?: React.FC<UserInputFormInputPropTypes> 
      }
      export const defaultReactFormContext: GQLReactFormContext = {
        form: 'form' as any,
        div: ({...props}: StandardProps) => <div {...props} />,
        label: ({...props}: StandardProps) => <label {...props} />,
        labelTextWrapper: ({...props}: StandardProps) => <h4 {...props} />,
        button: ({...props}: StandardProps & {onClick?: (e?: {preventDefault: () => any}) => any}) => <button {...props} onClick={e => {e.preventDefault(); props.onClick?.()}}/>,
        listWrapper: ({...props}: StandardProps) => <ol {...props} />,
        listItem: ({...props}: StandardProps) => <li {...props} />,
        submitButton: ((props: {text: string}) => <input type=\\"submit\\" {...props} value={props.text} /> ),
        input: ((props) => {
          const {path, scalar, name} = props
          const ctx = React.useContext(GQLReactFormContext)
          const DivComponent = ctx.div || defaultReactFormContext.div
          const typeofValue = typeof props.value
            return (
            <DivComponent  path={path} scalar={scalar} name={name}>
          
            <label>
              <strong>{props.label}</strong><br />
              <input value={props.value} type={props.value === undefined
                ? \\"string\\"
                : typeofValue === 'string'
                ? 'text'
                : typeofValue === 'number'
                ? 'number'
                : ''
              } onChange={(e) =>
                props.onChange(e.target.value)} />
            </label>
            
          </DivComponent>)

          }),

      }
      export const GQLReactFormContext = React.createContext<Partial<GQLReactFormContext>>(defaultReactFormContext)

    
/**********************
 * Default Values
 * *******************/
  export const defaultUserInputScalar = {

          get id (): Scalars[\\"Int\\"]|undefined {
            
          return JSON.parse(JSON.stringify(0))
        
          },
    

          get name (): Scalars[\\"String\\"] {
            
          return JSON.parse(JSON.stringify(\\"\\"))
        
          },
    

          get email (): Scalars[\\"String\\"] {
            
          return JSON.parse(JSON.stringify(\\"\\"))
        
          },
    

          get password (): Scalars[\\"String\\"]|undefined {
            
          return JSON.parse(JSON.stringify(\\"\\"))
        
          },
    

          get mother (): UserInput|undefined {
            return undefined
          },
    

          get father (): UserInput|undefined {
            return undefined
          },
    

          get friends (): UserInput[] {
            
          return JSON.parse(JSON.stringify([]))
        
          },
    

          get followers (): UserInput[]|undefined {
            return undefined
          },
    
}
/**********************
 * Scalar Form Fragments
 * *******************/
  
    export interface IntFormInputPropTypes {
      optional: boolean,
      label: string,
      value?: Maybe<Scalars[\\"Int\\"]>,
      scalarName: string,
      name: string,
      parentPath: string,
      onChange: (value?: Scalars[\\"Int\\"]) => any
    }
        export const IntFormInput = React.memo((props: IntFormInputPropTypes) => {
          const {parentPath, label, name, value, onChange } = props
const scalar = \\"Int\\"
const path = [parentPath, name].join('.')
const ctx = React.useContext(GQLReactFormContext)
const DivComponent = ctx.div || defaultReactFormContext.div
const ButtonComponent = ctx.button || defaultReactFormContext.button
const LabelTextWrapperComponent = ctx.labelTextWrapper || defaultReactFormContext.labelTextWrapper
const InputComponent = ctx.input || defaultReactFormContext.input
          return (
        <InputComponent onChange={props.onChange as any} value={value === undefined || value===null? \\"\\" : value} label={label} path={path} scalar={scalar} name={name}>
          
          </InputComponent>)
        })
    


    export interface StringFormInputPropTypes {
      optional: boolean,
      label: string,
      value?: Maybe<Scalars[\\"String\\"]>,
      scalarName: string,
      name: string,
      parentPath: string,
      onChange: (value?: Scalars[\\"String\\"]) => any
    }
        export const StringFormInput = React.memo((props: StringFormInputPropTypes) => {
          const {parentPath, label, name, value, onChange } = props
const scalar = \\"String\\"
const path = [parentPath, name].join('.')
const ctx = React.useContext(GQLReactFormContext)
const DivComponent = ctx.div || defaultReactFormContext.div
const ButtonComponent = ctx.button || defaultReactFormContext.button
const LabelTextWrapperComponent = ctx.labelTextWrapper || defaultReactFormContext.labelTextWrapper
const InputComponent = ctx.input || defaultReactFormContext.input
          return (
        <InputComponent onChange={props.onChange as any} value={value === undefined || value===null? \\"\\" : value} label={label} path={path} scalar={scalar} name={name}>
          
          </InputComponent>)
        })
    


    export interface UserInputFormInputPropTypes {
      optional: boolean,
      label: string,
      value?: Maybe<UserInput>,
      scalarName: string,
      name: string,
      parentPath: string,
      onChange: (value?: UserInput) => any
    }
        export const UserInputFormInput = React.memo((props: UserInputFormInputPropTypes) => {
          const {parentPath, label, name, value, onChange } = props
const scalar = \\"UserInput\\"
const path = [parentPath, name].join('.')
const ctx = React.useContext(GQLReactFormContext)
const DivComponent = ctx.div || defaultReactFormContext.div
const ButtonComponent = ctx.button || defaultReactFormContext.button
const LabelTextWrapperComponent = ctx.labelTextWrapper || defaultReactFormContext.labelTextWrapper
          
      if(value === undefined || value === null ){
        return <DivComponent  path={path} scalar={scalar} name={name}>
          <ButtonComponent onClick={() => onChange(JSON.parse(JSON.stringify(defaultUserInputScalar)))
          } path={path} scalar={scalar} name={name}>
          Add {label}
          </ButtonComponent>
          </DivComponent>
      }
return <DivComponent className={\\"mutationFormNested\\"} path={path} scalar={scalar} name={name}>
          
            <LabelTextWrapperComponent  path={path} scalar={scalar} name={name}>
          {label}
          </LabelTextWrapperComponent>
            <UserInputFormInput value={value?.users === null? undefined : value?.users} scalarName={\\"UserInput\\"} name={\\"users\\"} optional={true} label={\\"Users\\"} parentPath={path} onChange={(newValue = JSON.parse(JSON.stringify(defaultUserInputScalar))) => onChange({...value, ['users']: newValue})} />
        
          </DivComponent>
        })
    


    export interface UserInputFormInputAsListPropTypes {
      optional: boolean,
      label: string,
      value?: Maybe<UserInput>[],
      scalarName: string,
      name: string,
      parentPath: string,
      onChange: (value?: UserInput[]) => any
    }
        export const UserInputFormInputAsList = React.memo((props: UserInputFormInputAsListPropTypes) => {
          const {parentPath, label, name, value, onChange } = props
const scalar = \\"UserInput\\"
const path = [parentPath, name].join('.')
const ctx = React.useContext(GQLReactFormContext)
const DivComponent = ctx.div || defaultReactFormContext.div
const ButtonComponent = ctx.button || defaultReactFormContext.button
const LabelTextWrapperComponent = ctx.labelTextWrapper || defaultReactFormContext.labelTextWrapper
const ListItemComponent = ctx.listItem || defaultReactFormContext.listItem
const valueMapRef = React.useRef<
          {id: string, value: Maybe<UserInput>}[]
        >((value||[]).map(v => ({id: uniqueId(\\"users\\"), value: v})))
const addItem=() => {
          valueMapRef.current = [...valueMapRef.current, {id: uniqueId('users'), value: JSON.parse(JSON.stringify(defaultUserInputScalar))} ];
          onChange(valueMapRef.current.map(i => i.value === null ? JSON.parse(JSON.stringify(defaultUserInputScalar)): i.value))
        }
const insertItem=(index: number) =>  {
            valueMapRef.current = [
              ...valueMapRef.current.slice(0, index),
              {id: uniqueId('users'), value: JSON.parse(JSON.stringify(defaultUserInputScalar))},
              ...valueMapRef.current.slice(index) ];
            onChange(valueMapRef.current.map(i => i.value === null ? JSON.parse(JSON.stringify(defaultUserInputScalar)): i.value))
        }
const removeItem=(index: number) => {
          valueMapRef.current = [...valueMapRef.current.slice(0, index), ...valueMapRef.current.slice(index+1) ]
          onChange(valueMapRef.current.map(i => i.value === null ? JSON.parse(JSON.stringify(defaultUserInputScalar)): i.value))
        }
const ListWrapperComponent = ctx.listWrapper || defaultReactFormContext.listWrapper
          return (
    <DivComponent className={\\"mutationFormNested mutationFormList\\"} path={path} scalar={scalar} name={name}>
          
    {label && <LabelTextWrapperComponent  path={path} scalar={scalar} name={name}>
          {label}
          </LabelTextWrapperComponent>}
    <ListWrapperComponent  path={path} scalar={scalar} name={name}>
          
        {valueMapRef.current.length > 0 ? (
          valueMapRef.current.map((item, index) => (
            <ListItemComponent key={item.id} path={path} scalar={scalar} name={name}>
          <ButtonComponent onClick={() => removeItem(index)} path={path} scalar={scalar} name={name}>
          X
          </ButtonComponent>

            <ButtonComponent onClick={() => insertItem(index)} path={path} scalar={scalar} name={name}>
          +
          </ButtonComponent>
            <UserInputFormInput optional={false} label={\\"\\"} value={item.value} scalarName={\\"UserInput\\"} name={String(index)} parentPath={path} onChange={(newValue = JSON.parse(JSON.stringify(defaultUserInputScalar))) => {
                    valueMapRef.current = valueMapRef.current.map(i => i.id === item.id ? {id: item.id, value: newValue} : i)
                    onChange(valueMapRef.current.map(i => i.value === null ?JSON.parse(JSON.stringify(defaultUserInputScalar)) : i.value))
                }} />
          </ListItemComponent>
          ))
        ) : (
          <ButtonComponent onClick={addItem} path={path} scalar={scalar} name={name}>
          +
          </ButtonComponent>
        )}
          </ListWrapperComponent>
      
          </DivComponent>
    )
        })
    
/***************************
* forms Forms
* *************************/
  

  export const addUsersDefaultValues = {
  users: []
  };


  export interface AddUsersFormVariables {
  users: UserInput[]
  }



  export const AddUsersForm = (
  {
    initialValues = addUsersDefaultValues,
    onSubmit,
    ...formProps} : React.DetailedHTMLProps<
  React.FormHTMLAttributes<HTMLFormElement>,
  HTMLFormElement
  > & { initialValues?: Partial<AddUsersFormVariables>, onSubmit: (values: AddUsersFormVariables)=> any}) => {
  const [value, setValue]= React.useState(initialValues || {})
  const ctx = React.useContext(GQLReactFormContext)
  const FormComponent = ctx.form || defaultReactFormContext.form
  const SubmitButtonComponent = ctx.submitButton || defaultReactFormContext.submitButton
  return (
      <FormComponent scalar=\\"\\" name=\\"\\" onSubmit={(e) => {
        e?.preventDefault?.()
        onSubmit(value as any)
      }} {...formProps} path=\\"\\">
        <UserInputFormInputAsList value={value?.users} label={\\"Users\\"} parentPath={\\"root\\"} onChange={(value) => {
                console.log('onChange users', value)
                setValue(oldVal => ({...oldVal, ['users']: value}))
              }} scalarName={\\"UserInput\\"} name={\\"users\\"} optional={false} />
        <SubmitButtonComponent text=\\"submit\\" />
      </FormComponent>
  )
  };
    
/***************************
 * MetaData Export
 * *************************/
  
  export const mutationsMetaData = [
  {
    \\"name\\": \\"addUsers\\",
    \\"variables\\": [
      {
        \\"accessChain\\": [],
        \\"endedFromCycle\\": false,
        \\"scalarName\\": \\"UserInput\\",
        \\"name\\": \\"users\\",
        \\"tsType\\": \\"UserInput\\",
        \\"optional\\": false,
        \\"asList\\": true,
        \\"children\\": [
          {
            \\"accessChain\\": [
              \\"UserInput\\"
            ],
            \\"endedFromCycle\\": false,
            \\"scalarName\\": \\"UserInput\\",
            \\"name\\": \\"users\\",
            \\"tsType\\": \\"UserInput\\",
            \\"optional\\": true,
            \\"asList\\": false,
            \\"children\\": [
              {
                \\"accessChain\\": [
                  \\"UserInput\\",
                  \\"Int\\"
                ],
                \\"endedFromCycle\\": false,
                \\"scalarName\\": \\"Int\\",
                \\"name\\": \\"id\\",
                \\"tsType\\": \\"Scalars[\\\\\\"Int\\\\\\"]\\",
                \\"optional\\": true,
                \\"asList\\": false,
                \\"children\\": null
              },
              {
                \\"accessChain\\": [
                  \\"UserInput\\",
                  \\"String\\"
                ],
                \\"endedFromCycle\\": false,
                \\"scalarName\\": \\"String\\",
                \\"name\\": \\"name\\",
                \\"tsType\\": \\"Scalars[\\\\\\"String\\\\\\"]\\",
                \\"optional\\": false,
                \\"asList\\": false,
                \\"children\\": null
              },
              {
                \\"accessChain\\": [
                  \\"UserInput\\",
                  \\"String\\"
                ],
                \\"endedFromCycle\\": false,
                \\"scalarName\\": \\"String\\",
                \\"name\\": \\"email\\",
                \\"tsType\\": \\"Scalars[\\\\\\"String\\\\\\"]\\",
                \\"optional\\": false,
                \\"asList\\": false,
                \\"children\\": null
              },
              {
                \\"accessChain\\": [
                  \\"UserInput\\",
                  \\"String\\"
                ],
                \\"endedFromCycle\\": false,
                \\"scalarName\\": \\"String\\",
                \\"name\\": \\"password\\",
                \\"tsType\\": \\"Scalars[\\\\\\"String\\\\\\"]\\",
                \\"optional\\": true,
                \\"asList\\": false,
                \\"children\\": null
              },
              {
                \\"accessChain\\": [
                  \\"UserInput\\",
                  \\"UserInput\\"
                ],
                \\"endedFromCycle\\": true,
                \\"scalarName\\": \\"UserInput\\",
                \\"name\\": \\"mother\\",
                \\"tsType\\": \\"UserInput\\",
                \\"optional\\": true,
                \\"asList\\": false,
                \\"children\\": null
              },
              {
                \\"accessChain\\": [
                  \\"UserInput\\",
                  \\"UserInput\\"
                ],
                \\"endedFromCycle\\": true,
                \\"scalarName\\": \\"UserInput\\",
                \\"name\\": \\"father\\",
                \\"tsType\\": \\"UserInput\\",
                \\"optional\\": true,
                \\"asList\\": false,
                \\"children\\": null
              },
              {
                \\"accessChain\\": [
                  \\"UserInput\\",
                  \\"UserInput\\"
                ],
                \\"endedFromCycle\\": true,
                \\"scalarName\\": \\"UserInput\\",
                \\"name\\": \\"friends\\",
                \\"tsType\\": \\"UserInput\\",
                \\"optional\\": false,
                \\"asList\\": true,
                \\"children\\": null
              },
              {
                \\"accessChain\\": [
                  \\"UserInput\\",
                  \\"UserInput\\"
                ],
                \\"endedFromCycle\\": true,
                \\"scalarName\\": \\"UserInput\\",
                \\"name\\": \\"followers\\",
                \\"tsType\\": \\"UserInput\\",
                \\"optional\\": true,
                \\"asList\\": true,
                \\"children\\": null
              }
            ]
          }
        ]
      }
    ]
  }
]
  
  "
`;

exports[`React Apollo Imports should import React and forms dependencies 1`] = `
Array [
  "import * as React from 'react';",
]
`;

exports[`React Apollo Imports should import React and forms dependencies 2`] = `
"
    /**************************
     * utilities
     *************************/
    let idNonce = 0;
    const uniqueId = (inStr: string) => inStr+(idNonce++)
    

    
      interface StandardProps {
        children?: React.ReactNode
        path: string
        id?: string
        className?: string
        name: string
        scalar: string
      }
      interface FormPrimeInput extends React.FC<{onChange: (value: number|string) => number|string, value?: number|string, label: string} & StandardProps> {}
      interface GQLFormStandardComponent<T = {}> extends React.FC<StandardProps & T> { }
      export interface GQLReactFormContext {
        form: GQLFormStandardComponent< {onSubmit: (e?: {preventDefault?: () => any}) => any} >,
        div: GQLFormStandardComponent,
        label: GQLFormStandardComponent,
        labelTextWrapper: GQLFormStandardComponent,
        button: GQLFormStandardComponent<{onClick?: (e?: {preventDefault?: () => any}) => any }>,
        listWrapper: GQLFormStandardComponent,
        listItem: GQLFormStandardComponent,
        submitButton: React.FC<{text: string}>
        input: FormPrimeInput,
        
      }
      export const defaultReactFormContext: GQLReactFormContext = {
        form: 'form' as any,
        div: ({...props}: StandardProps) => <div {...props} />,
        label: ({...props}: StandardProps) => <label {...props} />,
        labelTextWrapper: ({...props}: StandardProps) => <h4 {...props} />,
        button: ({...props}: StandardProps & {onClick?: (e?: {preventDefault: () => any}) => any}) => <button {...props} onClick={e => {e.preventDefault(); props.onClick?.()}}/>,
        listWrapper: ({...props}: StandardProps) => <ol {...props} />,
        listItem: ({...props}: StandardProps) => <li {...props} />,
        submitButton: ((props: {text: string}) => <input type=\\"submit\\" {...props} value={props.text} /> ),
        input: ((props) => {
          const {path, scalar, name} = props
          const ctx = React.useContext(GQLReactFormContext)
          const DivComponent = ctx.div || defaultReactFormContext.div
          const typeofValue = typeof props.value
            return (
            <DivComponent  path={path} scalar={scalar} name={name}>
          
            <label>
              <strong>{props.label}</strong><br />
              <input value={props.value} type={props.value === undefined
                ? \\"string\\"
                : typeofValue === 'string'
                ? 'text'
                : typeofValue === 'number'
                ? 'number'
                : ''
              } onChange={(e) =>
                props.onChange(e.target.value)} />
            </label>
            
          </DivComponent>)

          }),

      }
      export const GQLReactFormContext = React.createContext<Partial<GQLReactFormContext>>(defaultReactFormContext)

    
/**********************
 * Default Values
 * *******************/
  
/**********************
 * Scalar Form Fragments
 * *******************/
  
/***************************
* forms Forms
* *************************/
  
/***************************
 * MetaData Export
 * *************************/
  
  export const mutationsMetaData = []
  
  "
`;

exports[`React Apollo Imports should import React and forms dependencies 3`] = `undefined`;
